stackable-certs/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+        source: x509_cert::der::Error,
+        key_encoding: KeyEncoding,
+    },
+
+    #[snafu(display("failed to serialize private key as PKCS8 {key_encoding}"))]
+    SerializePrivateKey {
+        source: p256::pkcs8::Error,
+        key_encoding: KeyEncoding,
+    },
+
+    #[snafu(display("failed to deserialize private key from PKCS8 {key_encoding}"))]
+    DeserializePrivateKey {
+        source: E,
+        key_encoding: KeyEncoding,
+    },
+
+    #[snafu(display("failed to write file"))]
+    WriteFile { source: std::io::Error },
+
+    #[snafu(display("failed to read file"))]
+    ReadFile { source: std::io::Error },
+}
+
+/// Contains the certificate and the signing / embedded key pair.
+///
+/// A [`CertificateAuthority`](crate::ca::CertificateAuthority) uses this struct
+/// internally to store the signing key pair which is used to sign the CA
+/// itself (self-signed) and all child leaf certificates. Leaf certificates on
+/// the other hand use this to store the bound keypair.
+#[derive(Debug)]
+pub struct CertificatePair<S>
+where
+    S: KeypairExt,
+    <S::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    certificate: Certificate,
+    key_pair: S,
+}
+
+impl<S> CertificatePair<S>
+where
+    S: KeypairExt,
+    <S::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    /// Returns a reference to the [`Certificate`].
+    pub fn certificate(&self) -> &Certificate {
+        &self.certificate
+    }
+
+    /// Returns a reference to the (signing) key pair.
+    pub fn key_pair(&self) -> &S {
+        &self.key_pair
+    }
+}
+
+#[cfg(feature = "rustls")]
+impl<S> CertificatePair<S>
+where
+    S: KeypairExt + 'static,
+    <S::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    pub fn certificate_der(
+        &self,
+    ) -> Result<CertificateDer<'static>, CertificatePairError<S::Error>> {
+        let der = self
+            .certificate
+            .to_der()
+            .context(SerializeCertificateSnafu {
+                key_encoding: KeyEncoding::Der,
+            })?
+            .into();
+
+        Ok(der)
+    }
+
+    pub fn private_key_der(
+        &self,
+    ) -> Result<PrivateKeyDer<'static>, CertificatePairError<S::Error>> {
+        // FIXME (@Techassi): Can we make this more elegant?
+        let doc = self
+            .key_pair
+            .signing_key()
+            .to_pkcs8_der()
+            .context(SerializePrivateKeySnafu {
+                key_encoding: KeyEncoding::Der,
+            })?;
+
+        let bytes = doc.to_bytes().deref().to_owned();
+        let der = PrivateKeyDer::from(PrivatePkcs8KeyDer::from(bytes));
+
+        Ok(der)
+    }
+}
+
+/// Provides functions to work with CAs stored in Kubernetes secrets.
+///
+/// Namely these function enable:
+///
+/// - decoding a certificate from a Kubernetes secret
+/// - encoding a certificate as a Kubernetes secret
+#[cfg(feature = "k8s")]
+pub trait CertificatePairExt: Sized {
```

Comment:
This still isn't an extension trait in the traditional sense. I'm not sure it should be a trait at all, but if it is then it should have a better name.

Reply:
I could instead implement the function directly on the struct and use the feature gates on the functions directly.

Reply:
Changed in a372b7a

Reply:
They can also be freestanding functions, this isn't Java. :P

Reply:
I know haha :) But I like the style `CertificateAuthority::from_secret()`